### PR TITLE
fix: fail when input type has `@protected` directive

### LINKED ISSUE #1596: bug: `@protected` directive should not be allowed on input types
Currently, there are no errors thrown when we use a `@protected` annotation on an input type. Ideally it should not be allowed.

```graphql
schema @upstream(baseURL: "http://jsonplaceholder.typicode.com") {
  query: Query
}

type Query {
  posts: [Post] @http(path: "/posts")
}

type Mutation {
  newPost(post: NewPost): Post @http(path: "/posts", method: POST, body: "{{args.post}}")
}

input NewPost @protected {
  name: String
}

type Post {
  id: Int!
  userId: Int!
  title: String!
  body: String!
}
```

```bash
 cargo run -- check  ./examples/auth.graphql
 INFO Env file: "/Users/tushar/Documents/Projects/tailcall/.env" loaded
 INFO File read: ./examples/auth.graphql ... ok
 INFO Config ./examples/auth.graphql ... ok
 INFO N + 1 detected: 0
```

**Expected Error**
```
Input types can not be protected [@ NewPost]
```

### PR DESCRIPTION
**Pull request by ologbonowiwi**
- **test: write `.md` test with input `@protected`**
- **test: include original example types**
- **test: load protected directive for input type**
- **feat: add flag to check whether type is input or not**
- **fix: fail when type is input**
- **test: update snapshots**

**Issue Reference(s):**  
Fixes #1596
/claim #1596

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [x] I have performed a self-review of my code.
- [x] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

## Summary by CodeRabbit

- **New Features**
	- Introduced a required "implements" property in GraphQL type schema definitions.
	- Added a "input" boolean property to GraphQL types to indicate if the type is an input.
	- Enhanced validation for protected input types in GraphQL schemas.

- **Bug Fixes**
	- Adjusted the control flow in the update function for protected fields to include an `object_name` argument, improving the accuracy of updates.

- **Tests**
	- Added a new test case to verify the handling and error reporting of protected input types in GraphQL queries and mutations.

### CODE REVIEW

File: src/blueprint/operators/protected.rs [hunk 1]
```diff
@@ -10,6 +10,10 @@ pub fn update_protected<'a>(
     TryFold::<(&ConfigModule, &Field, &config::Type, &'a str), FieldDefinition, String>::new(
         |(config, field, type_, _), mut b_field| {
             if field.protected.is_some() || type_.protected.is_some() {
+                if type_.input {
+                    return Valid::fail("Input types can not be protected".to_owned());
```
Comment (ologbonowiwi) [inline src/blueprint/operators/protected.rs:14]:
```suggestion
                    return Valid::fail("@protected can't be used on input types".to_owned());
```
this sounds more natural to me, let me know what you think @tusharmath

File: tests/snapshots/execution_spec__input-type-protected-error.md_errors.snap [hunk 2]
```diff
@@ -0,0 +1,24 @@
+---
+source: tests/execution_spec.rs
+expression: errors
+---
+[
+  {
+    "message": "Input types can not be protected",
```
Comment (ologbonowiwi) [inline tests/snapshots/execution_spec__input-type-protected-error.md_errors.snap:7]:
```suggestion
    "message": "@protected can't be used on input types",
```

File: tests/snapshots/execution_spec__input-type-protected-error.md_errors.snap [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    "trace": [
+      "Input",
+      "name",
+      "@protected"
+    ],
+    "description": null
+  },
+  {
+    "message": "Input types can not be protected",
```
Comment (ologbonowiwi) [inline tests/snapshots/execution_spec__input-type-protected-error.md_errors.snap:16]:
```suggestion
    "message": "@protected can't be used on input types",
```

**Review by ologbonowiwi — COMMENTED** [review]
in case the other suggestion gets merged.

File: tests/snapshots/execution_spec__input-type-protected-error.md_errors.snap [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+source: tests/execution_spec.rs
+expression: errors
+---
+[
+  {
+    "message": "Input types can not be protected",
+    "trace": [
+      "Input",
+      "name",
+      "@protected"
```
Comment (tusharmath) [inline tests/snapshots/execution_spec__input-type-protected-error.md_errors.snap:11]:
```suggestion
      "@protected"
```

Comment (ologbonowiwi) [reply tests/snapshots/execution_spec__input-type-protected-error.md_errors.snap:11]: Issue created to address this: https://github.com/tailcallhq/tailcall/issues/1599

File: src/blueprint/operators/protected.rs [hunk 5]
```diff
@@ -10,6 +11,20 @@ pub fn update_protected<'a>(
     TryFold::<(&ConfigModule, &Field, &config::Type, &'a str), FieldDefinition, String>::new(
         |(config, field, type_, _), mut b_field| {
             if field.protected.is_some() || type_.protected.is_some() {
+                let is_input = config.types.iter().any(|(_, _type)| {
+                    !_type.interface
+                        && _type.fields.iter().any(|(_, field)| {
+                            field.args.iter().any(|(_, arg)| {
+                                !scalar::is_scalar(&arg.type_of)
+                                    && config.find_type(&arg.type_of).is_some()
```
Comment (tusharmath) [inline src/blueprint/operators/protected.rs:19]: Input types information is pre computed in a hashmap. Check input_types method on config. Use it check

**Review by tusharmath — CHANGES_REQUESTED** [review]

File: src/blueprint/operators/protected.rs (hunk 5 above)
Comment (ologbonowiwi) [reply src/blueprint/operators/protected.rs]: Created a helper on `Type` that uses `input_types` for maximum reusability.

Comment (ologbonowiwi) [reply src/blueprint/operators/protected.rs]: As we do not use it anywhere else, and I don't see any other methods on `config.rs` receiving a `Config`, I removed it from there to keep the standard and created the `is_input` method on `protected.rs` file.

Feel free to use whatever version seems most convenient (the non-standard one is implemented on [16296d4](https://github.com/tailcallhq/tailcall/pull/1597/commits/16296d471d19b6f68a33b19962b53f4716ebd342))

**Comment by ologbonowiwi** [comment]
Do you have a suggestion to accomplish this differently @tusharmath? I wasn't able to found the type name anywhere else, so I needed to search for them on `config.types`. Let me know if there's anywhere else I can found this value.

File: src/config/from_document.rs [hunk 6]
```diff
@@ -237,9 +239,13 @@ fn to_enum(enum_type: EnumType) -> config::Type {
...
 }
-fn to_input_object(input_object_type: InputObjectType) -> Valid<config::Type, String> {
+fn to_input_object(
+    input_object_type: InputObjectType,
+    directives: &[Positioned<ConstDirective>],
+) -> Valid<config::Type, String> {
     to_input_object_fields(&input_object_type.fields)
-        .map(|fields| config::Type { fields, ..Default::default() })
+        .fuse(Protected::from_directives(directives.iter()))
+        .map(|(fields, protected)| config::Type { fields, protected, ..Default::default() })
```
Comment (ologbonowiwi) [inline src/config/from_document.rs:248]: Needed otherwise we would not have access to `protected` directive on `update_protected` for `input` types.

**Review by tusharmath — APPROVED** [review]

**Comment by tusharmath** [comment]
thanks @ologbonowiwi

**Pull request merged**